GoogleForCreators/web-stories-wp
# Hotlinking: Fix CORS Proxy issue and link validation bug

## Context



There was an issue with the CORS proxy not returning the correct response code, as well as an issue with URLs not being encoded before validated on the client-side.

Also, insertion when using the proxy was potentially quite slow, which was not very clear to users.

## Summary



This PR contains a few (significant) changes to improve the overall UX and performance of hotlinking.

1. **Accessibility**: The hotlinking dialog can now be submitted by pressing enter.
2. **User Feedback**: The "Submit" button is now disabled while insertion happens.  
  A spinner could perhaps be added in the future.
3. **Performance**: Overhaul insertion logic to reduce the number of times the video is fetched.  
  Mostly relevant when loading larger video files via the CORS proxy.

Regarding the latter, a video is now _preloaded_ only once, so that dimensions, length, and audio information can be retrieved.

If the user can upload media, the video is _seeked_ to the initial frame to generate a poster image, even _before_ the video is actually inserted.

Doing that _before_ insertion has two main reasons:

1. Doing it _after_ and updating the video using `updateElementsByResourceId` would cause an unexpected history entry.
2. Using `updateElementsByResourceId` for that would also mean the video gets re-rendered, which means playback stops.

## Relevant Technical Choices



Refactored a lot of the logic for preloading videos and extracting metadata, to reduce repetition in multiple areas.

## To-do



N/A

## User-facing changes



* The hotlinking dialog can now be submitted by pressing enter
* The "Submit" button is now disabled while insertion happens

## Testing Instructions






1.  Verify that external videos and images can be inserted via hotlinking
2. Verify that URLs like `https://foo` are rejected
2. Verify that the "Insert" button changes to "Inserting" while insertion is in progress

Some possible test files:

* https://c.tenor.com/Mc-DPcAOaNsAAAPo/lol-omg.mp4
* https://sample-videos.com/video123/mp4/720/big_buck_bunny_720p_30mb.mp4 (large)
* https://spacedmonkey.ams3.cdn.digitaloceanspaces.com/uploads/2021/10/sample_640%C3%97426.jpeg
* https://spacedmonkey.ams3.cdn.digitaloceanspaces.com/uploads/2021/10/sample-30s-1.webm (large)
* https://spacedmonkey.ams3.cdn.digitaloceanspaces.com/uploads/2021/10/sample_960x400_ocean_with_audio.mp4 (large)
* https://stream.mux.com/r4lT81CEvYAIutD01Zp7KAZAMnk2dbafb/high.mp4?token=eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6InQ5UHZucm9ZY0hQNjhYSmlRQnRHTEVVSkVSSXJ0UXhKIn0.eyJleHAiOjE2Mzc1OTUyNTcsImF1ZCI6InYiLCJzdWIiOiJyNGxUODFDRXZZQUl1dEQwMVpwN0tBWkFNbmsyZGJhZmIifQ.GedMsn0dzJHhNl7nj6jKpMgc7G4u7-hOUt98HBeiI5OAiiGw5_cWPBFMqscaUo7WA9MrNDpMVk6aSsXOvtGlaWvj58Tjh5C08Z1Wjabioba2zMN8MED-ZfYoqj5KOIkrDYV2hY0ZzT83LBCwL83uLbYCK2xtjvjFRUBsMimQtksLxVy6KN28x-ahHmexOwh-u5zdc3I9eEIRGi1x56KOWnOavOZVtJREiBZvZs7ApCvxGvydit5IoMUleJxia_TtitZHBbuLV-SHqAPyLc5-gYKYVRnFP3Z4brHHF2nkSerdA3EfMw6Dpcj85UjYzlk875rih9VNtqikPDdw0ki4_A

**Notes**

* Hotlinking large media files can be slow to insert because they're loaded through the WP site (proxy)

## Reviews

### Does this PR have a security-related impact?



No

### Does this PR change what data or activity we track or use?



No

### Does this PR have a legal-related impact?



No

## Checklist




---



Fixes #9696
